feat: Add support for volumes and volumeMounts

naimadswdn opened the pull request:
<!--
    Please read https://github.com/OT-CONTAINER-KIT/redis-operator/blob/main/CONTRIBUTING.md before submitting
    your pull request. Please fill in each section below to help us better prioritize your pull request. Thanks!
-->

**Description**

<!-- Please provide a summary of the change here. -->

<!-- Please link to all GitHub issue that this pull request implements(i.e. Fixes #123) -->
New feature added: support for volumes and volumeMounts for the all of the Redis deployment types.

**Type of change**

<!-- Please delete options that are not relevant. -->

* New feature (non-breaking change which adds functionality)

**Checklist**

- [X] Tests have been added/modified and all tests pass.
- [X] Functionality/bugs have been confirmed to be unchanged or fixed.
- [X] I have performed a self-review of my own code.
- [X] Documentation has been updated or added where necessary.

**Additional Context**

I need the volume and volumeMounts support to mount the external ACL file. 

I tested the solution on my own cluster (AKS) and it works as expected:
<img width="522" height="331" alt="image" src="https://github.com/user-attachments/assets/57964931-41fa-4d20-8f08-d400d97c3c35" />
<img width="595" height="65" alt="image" src="https://github.com/user-attachments/assets/0bf989f4-322a-4c00-b7e0-5e8d681a21bf" />

naimadswdn commented:
@shubham-cmyk @drivebyer @iamabhishek-dubey can I ask one of you to approve the workflow? I fixed all the things that failed recently.

naimadswdn commented:
@drivebyer right, I just spitted the PR. 
There is a new one for the PDB fix: https://github.com/OT-CONTAINER-KIT/redis-operator/pull/1563

naimadswdn commented:
@shubham-cmyk @drivebyer @iamabhishek-dubey Hi guys, the PR passed all the CI steps, can we merge it? :)

drivebyer commented:
If you’d like to add ACL configuration to your Redis cluster, you can refer to the examples here: [v1beta2/acl_config](https://github.com/OT-CONTAINER-KIT/redis-operator/tree/main/example/v1beta2/acl_config)

naimadswdn commented:
> If you’d like to add ACL configuration to your Redis cluster, you can refer to the examples here: [v1beta2/acl_config](https://github.com/OT-CONTAINER-KIT/redis-operator/tree/main/example/v1beta2/acl_config)

You are right, but I have a hard requirement to persist my ACL between Redis installation and k8s secret is not the best to store such data. 
That is why I am provisioning a separate StorageClass with the Retain policy and mount the PVC to the Redis Cluster components. 
This way I have a single source of truth, that persist between installations.

drivebyer commented:
> This way I have a single source of truth, that persist between installations.

The ACL data is stored in a dedicated Secret resource, making it independent from the Redis cluster lifecycle.

naimadswdn commented:
In my setup, I need to support the multitenancy solution with the single Redis Cluster. That means, there is a lot of ACL modifications and relying the k8s secret for that is not the best practice. 
Using the separate PVC indicates better performance and allows to track the changes.

I am not saying storing ACL in a secret is bad idea - it is good, but not for all situations. 
I need a multitenancy ACL support that dynamically updates and persist even if the Redis Helm chart was uninstalled and installed again. 

Basically this additional volumes can be used also for other configuration items in the future. ACL is just an example that I am relying on my project.

drivebyer commented:
Got it. That might make the setup a bit more complicated for users. Since we already support spec.acl, what do you think about adding the ACL volume configuration there?

drivebyer commented:
> I could but I wonder if that is not a step back from the current implementation. Right now, you can use the additionalVolumes that are part of this PR to attach not only the external ACL, but also for example:
> 
> * persistent volume for Redis, when you want to kept the data (AOF, RDB snapshots)
> * attach custom scripts from configMap,
> * attach cutom secrets from k8s secrets,
> * use some epheral volumes for caching or some dev/test environments,
> * use CSI volume like Azure Disk to store some data (ACL for example)
> * volume sharing between all Redis leaders or followers.

This PR definitely adds flexibility — but it also increases complexity for users, since they need to understand how to configure volume mounts.
For example, the ACL volume path is currently hardcoded as /etc/redis/user.acl.

That’s why I’d suggest keeping it simple and focusing on solving your specific requirement directly.
Maybe something like:
```
spec:
  acl:
    volumeSource:
      ...
```

Then the operator could automatically handle the volume mount for the Redis cluster, which would make it more user-friendly overall.

drivebyer commented on api/common/v1beta2/common_types.go:
```diff
@@ -15,6 +17,8 @@ type KubernetesConfig struct {
 	ImagePullSecrets                     *[]corev1.LocalObjectReference                          `json:"imagePullSecrets,omitempty"`
 	UpdateStrategy                       appsv1.StatefulSetUpdateStrategy                        `json:"updateStrategy,omitempty"`
 	PersistentVolumeClaimRetentionPolicy *appsv1.StatefulSetPersistentVolumeClaimRetentionPolicy `json:"persistentVolumeClaimRetentionPolicy,omitempty"`
+	AdditionalVolumes                    []corev1.Volume                                         `json:"additionalVolumes,omitempty"`
```
remove AdditionalVolumes/AdditionalVolumeMounts and related codes. or raise another pr to implement ACLConfig.PersistentVolumeClaim :)

Commit:
feat: Add support for volumes and volumeMounts

Add support for volumes and volumeMounts.
Cover the changes with the tests.
Update helm charts with the Volumes and VolumeMounts options.
Add missing helm chart support for persistentVolumeClaimRetentionPolicy.

Signed-off-by: Damian Seredyn <s-DSeredyn@aras.com>
